fdmi/plugins/fdmi_sample_plugin.c

```diff
@@ -0,0 +1,426 @@
+				     LAMBDA(void, (const char *string) {
+						     params->spp_local_addr = (char*)string;
+					     })),
+			M0_STRINGARG('h', "HA address",
+				     LAMBDA(void, (const char *str) {
+						     params->spp_hare_addr = (char*)str;
+					     })),
+			M0_STRINGARG('f', "Process FID",
+				     LAMBDA(void, (const char *str) {
+						     params->spp_process_fid = (char*)str;
+					     })),
+			M0_STRINGARG('p', "Profile options for client",
+				     LAMBDA(void, (const char *str) {
+						     params->spp_profile_fid = (char*)str;
+					     })),
+			M0_STRINGARG('g', "FDMI plugin fid",
+				     LAMBDA(void, (const char *str) {
+						     params->spp_fdmi_plugin_fid_s =
+							     (char*)str;
+					     })));
+	if (rc != 0)
+		return M0_ERR(rc);
+
+        /* All mandatory params must be defined. */
+	if (rc == 0 &&
+	    (params->spp_local_addr == NULL || params->spp_hare_addr == NULL ||
+	     params->spp_profile_fid == NULL || params->spp_process_fid == NULL ||
+	     params->spp_fdmi_plugin_fid_s == NULL)) {
+		fsp_usage();
+		return M0_ERR(-EINVAL);
+	}
+
+	rc = m0_fid_sscanf(params->spp_fdmi_plugin_fid_s, &params->spp_fdmi_plugin_fid);
+	if (rc != 0) {
+		rc = M0_ERR_INFO(rc, "Invalid FDMI plugin fid format: fid=%s",
+				 params->spp_fdmi_plugin_fid_s);
+	}
+
+	return rc;
+}
+
+/**
+ * Callback function called by the FDMI service to deliver the FDMI event
+ * to the plugin program.
+ */
+static int process_fdmi_record(struct m0_uint128 *rec_id,
+			       struct m0_buf fdmi_rec,
+		   	       struct m0_fid filter_id)
+{
+	struct m0_fol_rec fol_rec;
+
+	m0_fol_rec_init(&fol_rec, NULL);
+	m0_fol_rec_decode(&fol_rec, &fdmi_rec);
+	m0_dump_m0_fol_rec_to_json(&fol_rec);
+	m0_fol_rec_fini(&fol_rec);
+	return 0;
+}
+
+static int fdmi_service_start(struct m0_client *m0c)
+{
+	struct m0_reqh *reqh = &m0c->m0c_reqh;
+	struct m0_reqh_service_type *stype;
+	bool start_service = false;
+	int rc = 0;
+
+	stype = m0_reqh_service_type_find("M0_CST_FDMI");
+	if (stype == NULL) {
+		M0_LOG(M0_ERROR, "FDMI service type is not found.");
+		return M0_ERR_INFO(-EINVAL, "Unknown reqh service type: M0_CST_FDMI");
+	}
+
+	fsp_fdmi_service = m0_reqh_service_find(stype, reqh);
+	if (fsp_fdmi_service == NULL) {
+		rc = m0_reqh_service_allocate(&fsp_fdmi_service, &m0_fdmi_service_type, NULL);
+		if (rc != 0)
+			return M0_RC_INFO(rc, "Failed to allocate FDMI service.");
+		m0_reqh_service_init(fsp_fdmi_service, reqh, NULL);
+		start_service = true;
+	}
+
+	if (start_service)
+        	rc = m0_reqh_service_start(fsp_fdmi_service);
+	return M0_RC(rc);
+}
+
+static void fdmi_service_stop(void)
+{
+	if (fsp_fdmi_service != NULL) {
+		m0_reqh_service_stop(fsp_fdmi_service);
+		fsp_fdmi_service = NULL;
+	}
+}
+
+static int init_fdmi_plugin(struct m0_fsp_params *params)
+{
+	int rc;
+
+	fsp_pdo = m0_fdmi_plugin_dock_api_get();
+	const struct m0_fdmi_filter_desc fd;
+
+	const static struct m0_fdmi_plugin_ops pcb = {
+		.po_fdmi_rec = process_fdmi_record
+	};
```

Comment:
Variable definition should go first in the function.
Please move them forward.

Reply:
done